crates/ignore/src/dir.rs

```diff
@@ -507,6 +516,86 @@ impl Ignore {
     fn absolute_base(&self) -> Option<&Path> {
         self.0.absolute_base.as_ref().map(|p| &***p)
     }
+
+    pub fn get_match_metadata<'self_>(
+        &'self_ self,
+        match_metadata_token: MatchMetadataToken,
```

Comment:
As discussed, I for whatever reason really wanted to try and avoid doing allocations as part of this API extension

So the basic pattern is that internally a "match metadata token" gets passed around that "encodes" the indices (into an `Ignore`'s `Types`'s `selections` `Vec` I believe) of the `FileTypeDef`'s that a matching file corresponded to

And then here we "resolve" that token -> an iterator of references (attached to the `Ignore`'s lifetime I guess) to those `FileTypesDef`'s